Custom Auth Query Params - DO NOT MERGE

squid808 opened the pull request:
Add ability for users to include their own custom query parameters when making an authorization URL.

By providing a ```KeyValuePair<string, string>[]``` for the ```UserDefinedQueryParams``` member of the ```GoogleAuthorizationCodeFlow.Initializer```, this collection will be used to update the resulting url via ```Google.Apis.Requests.Parameters.ParameterUtils.IterateParameters()```.

An attempt at a unit test has been provided with ```/GoogleApis.Tests/Apis/Requests/Parameters/ParameterUtilsTest.cs```.

As this code was written between version changes to the base code, I tried to rebase the code to the upstream master (at the time) which may have cause some issues. As such, this pull request has been marked as DO NOT MERGE while everything can be checked and verified.

chrsmith commented:
> Add ability for users to include their own custom query parameters when making an authorization URL.

Why is this necessary? The authorization request only goes to accounts.google.com, right? Is there anything you need to specify via URL query parameters that isn't properly supported in the library? i.e. what does this PR unblock?

I'm not trying to be critical, I'm just trying to better understand the change.

squid808 commented:
No problem, thanks for asking. It was in response to [issue 638](https://github.com/google/google-api-dotnet-client/issues/638) at request of @peleyal .

chrsmith commented:
Gotcha, sorry for the confusion on our end. I'm trying to slowly ramp up in helping @peleyal out.

While it would probably be nicer to "properly support" features like the `login_hint` parameter by adding it as an optional field to the request object, having a general extensibility hook may come in handy in the future.

Should @peleyal or I go ahead and start to review the code, or do you still need to work out merge conflicts in your branch?

squid808 commented:
Understood, I look forward to working with you! So, as far as I know it should be good to go. Unfortunately things changed while I was working on it so I attempted to rebase it once, but my Git isn't super great yet so I'm not sure if I mucked anything up when I did that (my git client added in some errant lines in to the code when I rebased and merged, so I had to clean that up for instance). I think I cleaned all of that up, but I didn't rebase it off of the second update that came through towards the end, so it's not in sync with the upstream master of the Google code - not that it seems to affect any files that were changed in that time period.

Beyond that, the code seems to be working on my end. I also wrote the unit test, which passes, but that's assuming I'm testing things properly. You know what they say about the same person writing the unit tests and the code. =)

Please feel free to review and let me know what I can do to work things out - if for some reason I need to re-merge, rebase or just create a new fork and PR all together please don't hesitate to let me know.

peleyal commented on Src/GoogleApis.Auth/OAuth2/Flows/GoogleAuthorizationCodeFlow.cs:
```diff
@@ -39,6 +40,10 @@ public class GoogleAuthorizationCodeFlow : AuthorizationCodeFlow
+        private readonly KeyValuePair<string, string>[] userDefinedQueryParams;
+
+        /// <summary>Gets the user defined query parameters.</summary>
+        public KeyValuePair<string, string>[] UserDefinedQueryParams { get { return userDefinedQueryParams; } }
```
Why not having IEnumerable<KeyValuePair<string, string>>? In general I prefer IEnumerable over primitive array.

squid808 replied:
No problem, works for me. I've updated the changes for the next commit coming shortly.

peleyal commented on Src/GoogleApis.Auth/OAuth2/Requests/GoogleAuthorizationCodeRequestUrl.cs:
```diff
@@ -58,6 +59,13 @@ public class GoogleAuthorizationCodeRequestUrl : AuthorizationCodeRequestUrl
+        /// Gets or sets a collection of user defined query parameters to facilitate any not explicitly supported
+        /// by the library which will be included in the resultant authentication URL.
+        /// </summary>
+        [Google.Apis.Util.RequestParameterAttribute("user_defined_query_parameters",
```
1. I'm still not sure it's the right place to have it, I'll keep reviewing the code and update this comment.
2. I'm pretty sure that we don't need Google.Apis.Util.RequestParameterAttribute here, since it's not field recognized by the server. We only add this for field that the server knows, so we can have something like "user_defined_query_patameters=blahblahblah in the URL

peleyal replied:
I think that the thing that I don't like here is that we have a name for the RequestParameterAttribute. People might wonder what is this user_defined_... parameter?

I suggest adding another constructor to https://github.com/google/google-api-dotnet-client/blob/1dce119eb5200be57a6ff24159b895594d6bd558/Src/Support/GoogleApis.Core/Apis/Util/RequestParameterAttribute.cs without name, just type and use it here. What do you say?

peleyal commented on Src/GoogleApis.Core/Apis/Requests/Parameters/ParameterUtils.cs:
```diff
@@ -121,7 +121,17 @@ private static void IterateParameters(object request, Action<RequestParameterTyp
-                    action(attribute.Type, name, value);
+                    if (attribute.Type == RequestParameterType.UserDefiniedQueries)
+                    {
+                        foreach (var pair in (KeyValuePair<string, string>[])value)
```
It is worth to have a type validation check, and log if it isn't (I mean the cast for KeyValuePair array)

squid808 replied:
No problem, I'll wrap that in a type comparison. However for the logging (assuming you meant this), looking at the example of how it's set up in [AuthorizationCodeFlow](https://github.com/google/google-api-dotnet-client/blob/master/Src/GoogleApis.Auth/OAuth2/Flows/AuthorizationCodeFlow.cs#L44) and [MediaDownloader](https://github.com/google/google-api-dotnet-client/blob/master/Src/GoogleApis/Apis/%5BMedia%5D/Download/MediaDownloader.cs#L38) I would be unable to do this for ParameterUtils since the class is static.

If I try this:
```private static readonly ILogger Logger = ApplicationContext.Logger.ForType<ParameterUtils>();```

I get the error:
```Google.Apis.Requests.Parameters.ParameterUtils': static types cannot be used as type arguments```

Not sure how you'd want me to proceed on that one, so I'll hold off on that for now. Committing shortly.

peleyal commented on README.md:
```diff
@@ -1,5 +1,3 @@
-[![Build Status](https://travis-ci.org/google/google-api-dotnet-client.svg?branch=master)](https://travis-ci.org/google/google-api-dotnet-client)
```
I think that you removed it by mistake...  maybe as a result of bad merge :-1:  :)

squid808 replied:
Well crap. I've been having a lot of issues since I tried doing the rebasing, my self-taught is showing. Is there a way I can undo/ignore this? Alternately I've been considering just making a new fork and manually adding the changes below to it, there aren't that many and it could be done in decent time. I'm not sure if that's the 'proper' way to fix it, but... horseshoes and hand grenades this is not.

peleyal replied:
I'm sorry, I'm still not an expert on GitHub. My recommendation for you is to just revert those files, it should be easy. 
Because your CL is pretty small, you can also just create a new PR.
Up to you :+1:

peleyal commented on Src/GoogleApis.Core/Apis/Requests/Parameters/ParameterUtils.cs:
```diff
@@ -29,6 +30,9 @@ namespace Google.Apis.Requests.Parameters
     public static class ParameterUtils
     {
+        //TODO: Find out how to resolve the issue with setting up the Logger:
+        //private static readonly ILogger Logger = ApplicationContext.Logger.ForType<ParameterUtils>();
```
I think it should work if you add a static constructor and set the Logger there.
What is the exact error you get?

peleyal commented on Src/GoogleApis.Tests/Apis/Download/MediaDownloaderTest.cs:
```diff
@@ -28,10 +28,7 @@
 using NUnit.Framework;
 
 using Google.Apis.Download;
-using Google.Apis.Json;
```
Bad merge here as well.... and I'm pretty sure in the other files as well... Take a look in your diff and make sure you don't delete code...

peleyal commented:
A huge change was committed recently. I suggest you sync with head, and then create a new PR, it will be easier for you.
The main issue is that all code moved from Src to Src/Support, see #653 for more details.

squid808 commented:
I don't disagree. Unfortunately, this breaks things on my end as I don't have a new enough version of Visual Studio, which means that some (or all?) projects which require VS 14 (2015) in the new pull are crapping out on my end. A handful of the projects fail to load in the updated solution, and others are just seemingly empty. I'll work on troubleshooting this when I can, but until I can get it working I'll leave this here as a reference for just a bit longer.

squid808 commented:
Specifically, I'll be looking at [this section of the comments](https://github.com/google/google-api-dotnet-client/pull/653#commitcomment-15411486) more in detail when I can.

peleyal commented:
@mmdriley we shouldn't break people who are running on an old version of VS (like me...).
Maybe the new project configuration do break people, like @squid808 suggest here. It should be fixed.

mmdriley commented:
Sorry for the trouble. 

I'll test this on VS2013 next week. In the meantime, some projects _should_ appear mostly blank -- their source files were moved to shared projects.

Are you on the latest update (5, I think) of VS2013?

mmdriley commented:
Fix for VS2013 is coming in #660
